### PR TITLE
42Send only changed node metadata to clients instead of whole mapblock

### PR DESCRIPTION
**Pull request by est31**
Rebased version of  #3775 (and #3166).

Also added some of my own improvements. Will squash before merge.

### CODE REVIEW

File: src/server.cpp [hunk 1]
```diff
@@ -919,11 +921,20 @@ void Server::AsyncRunStep(bool initial_step)
...
-						prof.add("MEET_BLOCK_NODE_METADATA_CHANGED", 1);
-						setBlockNotSent(event->p);
+			case MEET_BLOCK_NODE_METADATA_CHANGED: {
+				verbosestream << "Server: MEET_BLOCK_NODE_METADATA_CHANGED" << std::endl;
+				prof.add("MEET_BLOCK_NODE_METADATA_CHANGED", 1);
+				// Don't send the change yet. Collect them to eliminate dupes.
+				node_meta_updates.remove(event->p);
+				node_meta_updates.push_back(event->p);
+
+				if (MapBlock *block = m_env->getMap().getBlockNoCreateNoEx(
```
Comment (rubenwardy) [inline src/server.cpp]: Could you split this into an assignment and a check?

File: src/server.cpp [hunk 2]
```diff
@@ -2123,6 +2136,61 @@ void Server::sendAddNode(v3s16 p, MapNode n, u16 ignore_id,
...
+	NodeMetadataList *meta_updates_list = new NodeMetadataList();
+
+	std::vector<u16> clients = m_clients.getClientIDs();
+	m_clients.lock();
+
+	for (std::vector<u16>::iterator i = clients.begin();
+			i != clients.end(); ++i) {
+		meta_updates_list->clear();
+		Player *player = m_env->getPlayer(*i);
+		RemoteClient* client = m_clients.lockedGetClientNoEx(*i);
```
Comment (rubenwardy) [inline src/server.cpp]: Star in the wrong place :P

File: src/server.cpp [hunk 3]
```diff
@@ -2123,6 +2136,61 @@ void Server::sendAddNode(v3s16 p, MapNode n, u16 ignore_id,
...
+	float maxd = far_d_nodes * BS;
+	NodeMetadataList *meta_updates_list = new NodeMetadataList();
+
+	std::vector<u16> clients = m_clients.getClientIDs();
+	m_clients.lock();
+
+	for (std::vector<u16>::iterator i = clients.begin();
+			i != clients.end(); ++i) {
+		meta_updates_list->clear();
+		Player *player = m_env->getPlayer(*i);
```
Comment (rubenwardy) [inline src/server.cpp]: Under what circumstances would player be null and client be non null?

Comment (rubenwardy) [inline src/server.cpp]: When the client is not emerged yet? If they're not emerged, will they have blocks stored that need modifying?

File: src/server.cpp (hunk 1 above)
Comment (est31) [inline src/server.cpp]: loool... it was that way before nrz demanded it to be this way.

File: src/server.cpp (hunk 3 above)
Comment (est31) [inline src/server.cpp]: I have no idea. In the past I had added an errorstream print to the else part of one of these checks, and it complained to me from time to time.

Removing the check here as you propose would possibly lead to the introduction of a hard to reproduce bug.

**Comment by Fixer-007** [comment]
Ready for test?

**Comment by est31** [comment]
Yes, its only about details.

**Comment by Fixer-007** [comment]
Failed to patch with the diff, networkprotocol.h hunk failed at 135.

**Comment by est31** [comment]
hrmm might have to rebase...

**Comment by est31** [comment]
Rebased.

File: src/server.cpp [hunk 4]
```diff
@@ -2126,6 +2139,61 @@ void Server::sendAddNode(v3s16 p, MapNode n, u16 ignore_id,
...
+			continue;
+		if (client->net_proto_version >= 28) {
+			for (std::list<v3s16>::const_iterator i2 = meta_updates.begin();
+					i2 != meta_updates.end(); ++i2) {
+				v3s16 pos = *i2;
+				NodeMetadata *meta = m_env->getMap().getNodeMetadata(pos);
+				if (!meta) {
+					continue;
+				}
+				if (player) {
```
Comment (rubenwardy) [inline src/server.cpp]: I was asking due to the if here - maybe the player check should be done before the inner loop. I don't know.

**Comment by everamzah** [comment]
It seems it's very fast.  Inventory to and from locked chests feels more responsive.

**Comment by Fixer-007** [comment]
Tried it with vending machines I see the differences, no more draw time spikes.

File: src/server.cpp (hunk 3 above)
Comment (nerzhul) [inline src/server.cpp]: exactly. Player object is kept in memory, not client.

File: src/network/clientpackethandler.cpp [hunk 5]
```diff
@@ -277,6 +277,32 @@ void Client::handleCommand_AddNode(NetworkPacket* pkt)
...
+
+void Client::handleCommand_NodemetaChanged(NetworkPacket *pkt)
+{
+	if (pkt->getSize() < 1)
+		return;
+
+	NodeMetadataList *meta_updates_list = new NodeMetadataList();
+	std::string datastring = pkt->readLongString();
+	std::istringstream is(datastring, std::ios::binary);
+	std::ostringstream oss(std::ios::binary);
```
Comment (ShadowNinja) [inline src/network/clientpackethandler.cpp]: Just use `stringstream` so you don't need all these copies.

**Comment by paramat** [comment]
How's this going? This really should be in release, i'll delay release if necessary.

File: src/server.cpp (hunk 4 above)
Comment (est31) [inline src/server.cpp]: It needs to be inside the inner loop unfortunately as it depends on pos. The only thing that can be put outside is the `v3f player_pos` calculation, but thats not worth it IMO.

**Comment by est31** [comment]
hmmm its spamming me with `WARNING[Main]: Map::setNodeMetadata(): Block not found` messages, no idea where they came from...

**Comment by Fixer-007** [comment]
Are you sure it is with 3848 only, or maybe it is spamming in default minetest? I've noticed them too but have not seen any effects.

**Comment by kilbith** [comment]
I noticed this message too, without this patch of course.

**Comment by Fixer-007** [comment]
I think I have server crash with this, I will need more time to confirm.

**Comment by paramat** [comment]
>  its spamming me with ```WARNING[Main]: Map::setNodeMetadata(): Block not found```

I get that without this patch sometimes, a message every few seconds.

**Comment by paramat** [comment]
>  its spamming me with WARNING[Main]: Map::setNodeMetadata(): Block not found messages

Sometimes i get that in master every few seconds.

**Comment by paramat** [comment]
Due to limited dev time, and because this can introduce subtle bugs, we have decided to work on this for longer and not rush it for this release.

**Comment by paramat** [comment]
From IRC, not possible for this release because:

hmmmm 	> no that's not going to be possible due to the protocol change

File: src/nodemetadata.cpp [hunk 6]
```diff
@@ -129,15 +135,19 @@ void NodeMetadataList::deSerialize(std::istream &is, IItemDefManager *item_def_m
...
+			p.X = readU16(is);
+			p.Y = readU16(is);
+			p.Z = readU16(is);
+		} else {
+			u16 p16 = readU16(is);
+			p.Z = p16 / MAP_BLOCKSIZE / MAP_BLOCKSIZE;
+			p16 &= MAP_BLOCKSIZE * MAP_BLOCKSIZE - 1;
+			p.Y = p16 / MAP_BLOCKSIZE;
+			p16 &= MAP_BLOCKSIZE - 1;
+			p.X = p16;
```
Comment (SmallJoker) [inline src/nodemetadata.cpp]: How about this one? Two divisions and no multiplication: (minimal improvement)
```C++
p.X = p16 & (MAP_BLOCKSIZE - 1);
p16 /= MAP_BLOCKSIZE;
p.Y = p16 & (MAP_BLOCKSIZE - 1);
p16 /= MAP_BLOCKSIZE;
p.Z = p16;
```

**Comment by est31** [comment]
Sorry I don't have time right now to work on this. I'd prefer if someone else took over.

**Comment by paramat** [comment]
@est31 could this be done for release?

File: src/nodemetadata.cpp (hunk 6 above)
Comment (sfan5) [inline src/nodemetadata.cpp]: I suggest not touching this code.

**Comment by Fixer-007** [comment]
Who wants some?

**Comment by paramat** [comment]
Adorable PR needs a loving home.

**Comment by SmallJoker** [comment]
Continued in #5268.

**Pull request closed without merging**